### DISCUSSION TITLE
OneDrive authorization error: Invalid request. Request is malformed or invalid.

### DISCUSSION THREAD

**Discussion by love2wind** [y-young/nazurin#61] [Q&A] [answered]
前面用heroku部署的时候，使用OneDrive是没有问题的，然后现在使用docker的时候OneDrive使用就出现下面的错误
```
OneDrive authorization error: AADSTS700082: The refresh token has expired due to inactivity. The token was issued on 2021-12-12T09:10:42.5991844Z and was inactive for 90.00:00:00.
Trace ID: 9a5a261d-07ed-4526-bd86-70ae42e20401
Correlation ID: 3c890b06-0b9d-4ff8-8ed4-4dd51fe6f9a0
Timestamp: 2022-11-23 19:04:26Z
```
我以为是refresh token过期了，重新获取refresh token后问题依旧，我又新注册了一个应用后，出现下面的错误
```
OneDrive authorization error: AADSTS700016: Application with identifier 'e7c5876a-518e-4194-a164-090f363e9b74' was not found in the directory 'niepan'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.
Trace ID: ee85a52b-3f75-4371-abe8-f12e8f783e01
Correlation ID: 5aac3a87-fe81-4274-87fb-e737acbdb963
Timestamp: 2022-11-24 03:48:30Z
```
我发现错误提示中的应用ID还是以前的，然后看了下Firestore数据库中的OneDrive数据还是原来的，就新建了一个数据库，重新部署后又出现了下面的错误
```
OneDrive authorization error: AADSTS9002313: Invalid request. Request is malformed or invalid.
Trace ID: 9a5a261d-07ed-4526-bd86-70ae07a01e01
Correlation ID: 1caf4017-8bf9-44fd-ba23-f1948b925f7e
Timestamp: 2022-11-24 08:00:21Z
```
目前停用了OneDrive存储，不知道上述问题是怎么回事？怎么解决？

**Comment by y-young** [comment, marked as answer]
检查一下OneDrive相关环境变量的设置，以及新创建的应用是否配置好了相关权限，具体参考[Wiki](https://github.com/y-young/nazurin/wiki/OneDrive)，此外建议把日志的上下文也贴出来

Update：**检查refresh token是否正确，使用ms-graph-token-tool时需注意在浏览器授权后还需要回到程序继续获取refresh token，重定向URL里的code其实并不是refresh token，只是用于获取refresh token的一个中间凭据，最终refresh token将显示在页面上**

**Reply by love2wind** [reply]
> 检查一下OneDrive相关环境变量的设置，以及新创建的应用是否配置好了相关权限，具体参考[Wiki](https://github.com/y-young/nazurin/wiki/OneDrive)，此外建议把日志的上下文也贴出来

OneDrive环境及应用权限没有问题。完全是按照wiki里的步骤做的，而且以前用heroku部署的时候是没有问题的。
我发现部署的时候有个错误提示，是不是这个问题造成的。
![image](https://user-images.githubusercontent.com/11996852/203952883-08f8d87c-f33f-49c6-8382-26645192fbf2.png)

**Reply by love2wind** [reply]
刚才搜索了一下，这个好像是PIP的问题
```
sudo pip install --upgrade pip
sudo pip install --upgrade urllib3
sudo pip install --upgrade requests
```
执行上述命令，docker-compose up -d后上述错误消失。但是OneDrive的问题依旧存在。

**Reply by love2wind** [reply]
> 检查一下OneDrive相关环境变量的设置，以及新创建的应用是否配置好了相关权限，具体参考[Wiki](https://github.com/y-young/nazurin/wiki/OneDrive)，此外建议把日志的上下文也贴出来

没找到日志文件在哪里？

**Reply by y-young** [reply]
> 没找到日志文件在哪里？

`docker logs nazurin_nazurin_1`

**Reply by love2wind** [reply]
> > 没找到日志文件在哪里？
> 
> `docker logs nazurin_nazurin_1`

![image](https://user-images.githubusercontent.com/11996852/204177821-b94cef02-3019-4d88-815e-5f7ab350d4d2.png)

**Reply by love2wind** [reply]
> > > 没找到日志文件在哪里？
> > 
> > 
> > `docker logs nazurin_nazurin_1`
> 
> ![image](https://user-images.githubusercontent.com/11996852/204177821-b94cef02-3019-4d88-815e-5f7ab350d4d2.png)

我又重新注册了应用做了一遍，还是依旧的错误，查看数据库还是没有OneDrive项
![image](https://user-images.githubusercontent.com/11996852/204182435-17a82dff-dc63-4db9-b4b5-142b7bc7c654.png)

**Reply by y-young** [reply]
看一下refresh token是否正确，如果refresh token粘贴不完整的话会报这个错

**Reply by love2wind** [reply]
> 看一下refresh token是否正确，如果refresh token粘贴不完整的话会报这个错

```
http://localhost:5000/getAToken?code=0.AXEAoGPn_LTfjEe1n09iTQ8bJfsvgRX7mzxNkfGIQTWGfwZxAOg.AgABAAIAAAD--DLA3VO7QrddgJg7WevrAgDs_wQA9P-c80UwSunv47eTfmMOqbwTa1RU3H8glkzDJ18R4-UovMV0xoBRlysBOO3jgWSFeMEDL0gLv8v9R_p9KOtjy1_DlbNtZJd3_KpEsKjz9gCB3EU6Pl5AHM7B9j5hUSYBOcXCC9Fzb6exX8jW6ONYfm1Bgh3gOjTfGBXFLPrUS8CWlTwNY0sUuUyyhyQuQKMK0gIfWsX5VHsj99q95eJ2vBG-AEwQMrQ6U7hIzQiJI3iJ0VjKXd-VlLsaatriTsnG-7IjVdCVEkQv7GtXbNAUeRDfPT9W0NyEkmDlFJZGj7C7BLeCG3PYSyhsEbIixecBonPXXn9nhkk86bFN-sqK6VtL7Ud7t-N-iKDlYqNv1AWmhSCXln1qe-w9QufVmmlzw_K1N1tQn4ZVTPsz_xluuNgwOPzAikfZ9jQiOpq95x43cLmUNw58TG0WKVinaHzqeORcDKBPP6bbDzUK2jUExMb5p0dPOpiIMFmcv4rixaF9HI_yvi84Nf2DK0712kxgscorsYJdilnvLLhdKoqpLU6aCVdyO2kAqRBeBRbZrKUPhQm3e0FRl_uNcu_G3AZoqJarGwTfgOwk4Yeg7C3J1ctRPfvadmJwklCuDbsIq32aDNnQzm1Od8yd-l07JPbCdBwLDL12MTxwGguwvZf3AbC6sk8cwSIvRuW158IGn4UKDryAhDijgchjCy0qg6DUN_2pn3AjALPHKvsD8W6LzugYKFD4LSLEN4Mnaxxz28RmckOnMC1XBhUbz8rRUma3AI7wJrYtqg7S7kZnMN69B3lUpA1x36N7_TIC-z_6fWFOXiS_w3_vTT8bRpwagVxOgppEFPr-CE3_tm_oZzYtpHCa7s4bLNLIeCyaLVn13fqx1Hx4Cea5jZ49DoeDKA&client_info=eyJ1aWQiOiIyZmU0NmMwMi05MTAwLTRmYzUtYjdkMS04MDM0ZDYwN2EwZDUiLCJ1dGlkIjoiZmNlNzYzYTAtZGZiNC00NzhjLWI1OWYtNGY2MjRkMGYxYjI1In0&state=ZokbgicRwBMdTUml&session_state=8e282efc-bd88-4f90-a8b3-643dcb9b3a11#
```
这是获取到的refresh token，直接填入是错误的，我对照以前获取的查看了一下，去掉头尾填入

```
0.AXEAoGPn_LTfjEe1n09iTQ8bJfsvgRX7mzxNkfGIQTWGfwZxAOg.AgABAAIAAAD--DLA3VO7QrddgJg7WevrAgDs_wQA9P-c80UwSunv47eTfmMOqbwTa1RU3H8glkzDJ18R4-UovMV0xoBRlysBOO3jgWSFeMEDL0gLv8v9R_p9KOtjy1_DlbNtZJd3_KpEsKjz9gCB3EU6Pl5AHM7B9j5hUSYBOcXCC9Fzb6exX8jW6ONYfm1Bgh3gOjTfGBXFLPrUS8CWlTwNY0sUuUyyhyQuQKMK0gIfWsX5VHsj99q95eJ2vBG-AEwQMrQ6U7hIzQiJI3iJ0VjKXd-VlLsaatriTsnG-7IjVdCVEkQv7GtXbNAUeRDfPT9W0NyEkmDlFJZGj7C7BLeCG3PYSyhsEbIixecBonPXXn9nhkk86bFN-sqK6VtL7Ud7t-N-iKDlYqNv1AWmhSCXln1qe-w9QufVmmlzw_K1N1tQn4ZVTPsz_xluuNgwOPzAikfZ9jQiOpq95x43cLmUNw58TG0WKVinaHzqeORcDKBPP6bbDzUK2jUExMb5p0dPOpiIMFmcv4rixaF9HI_yvi84Nf2DK0712kxgscorsYJdilnvLLhdKoqpLU6aCVdyO2kAqRBeBRbZrKUPhQm3e0FRl_uNcu_G3AZoqJarGwTfgOwk4Yeg7C3J1ctRPfvadmJwklCuDbsIq32aDNnQzm1Od8yd-l07JPbCdBwLDL12MTxwGguwvZf3AbC6sk8cwSIvRuW158IGn4UKDryAhDijgchjCy0qg6DUN_2pn3AjALPHKvsD8W6LzugYKFD4LSLEN4Mnaxxz28RmckOnMC1XBhUbz8rRUma3AI7wJrYtqg7S7kZnMN69B3lUpA1x36N7_TIC-z_6fWFOXiS_w3_vTT8bRpwagVxOgppEFPr-CE3_tm_oZzYtpHCa7s4bLNLIeCyaLVn13fqx1Hx4Cea5jZ49DoeDKA
```
还是错误的

**Reply by y-young** [reply]
Refresh token会直接显示在页面上，不需要手动从URL拆分，总之这个问题的原因应该是Refresh token不正确，可以尝试换一个client secret或换一个应用试一试

**Reply by love2wind** [reply]
> Refresh token会直接显示在页面上，不需要手动从URL拆分，总之这个问题的原因应该是Refresh token不正确，可以尝试换一个client secret或换一个应用试一试

搞定了，就是Refresh token的问题，原因是我前面一直是在服务器获取，这次我在本地获取，页面就显示了正确的Refresh token，重新部署后一切OK，感谢大佬耐心指导。。

**Reply by y-young** [reply]
> 原因是我前面一直是在服务器获取，这次我在本地获取，页面就显示了正确的Refresh token

原来如此，实际上在浏览器授权完还需要回到程序继续获取refresh token，刚才没注意到URL里的code其实并不是refresh token，只是用于获取refresh token的一个中间凭据

**Comment by love2wind** [comment]
非常感谢大佬的解答，问题圆满解决了
